src/controllers/Order.controller.ts

```diff
@@ -0,0 +1,222 @@
+import { checkLoggedIn } from '../untils/auth'
+import { User } from '../entities/User';
+import { Transactional, runOnTransactionCommit, runOnTransactionRollback } from 'typeorm-transactional';
+
+
+const orderRepository = AppDataSource.getRepository(Order);
+const orderDetailRepository = AppDataSource.getRepository(OrderDetail);
+const cartRepository = AppDataSource.getRepository(Cart);
+const cartItemRepository = AppDataSource.getRepository(CartItem);
+const productRepository = AppDataSource.getRepository(Product);
+
+const getUserCart = async (userId: number, res: Response) => {
+  const cart = await cartRepository.findOne({
+    where: { user: { id: userId } },
+  });
+  if (!cart) {
+    res.redirect('/');
+    return;
+  }
+  return cart;
+};
+// GET list item in cart
+export const getListItemOrder = async (
+  req: Request,
+  res: Response,
+  next: NextFunction,
+) => {
+  try {
+    const user = await checkLoggedIn(req, res);
+    const cart = await getUserCart(user.id, res);
+    const page = parseInt(req.query.page as string) || 1;
+    const offset = calculateOffset(page);
+    const [cartItems, totalItems] = await cartItemRepository.findAndCount({
+      where: { cart: { id: cart.id } },
+      relations: ['product'],
+      take: PAGE_SIZE,
+      skip: offset,
+    });
+    const totalPages = Math.ceil(totalItems / PAGE_SIZE);
+
+    // Get total price
+    const rawResult = await cartItemRepository
+      .createQueryBuilder('cartItem')
+      .select('count(cartItem.id) as items')
+      .addSelect('sum(cartItem.quantity * product.price) as total')
+      .leftJoin('cartItem.product', 'product')
+      .where('cartItem.cartId = :cartId', { cartId: cart.id })
+      .getRawOne();
+
+    const { total, items } = rawResult;
+
+    res.render('checkout', {
+      paginationItemsLinks: generatePaginationLinks(page, totalPages),
+      total,
+      items,
+      cartItems,
+    });
+  } catch (err) {
+    console.error(err);
+    next(err);
+  }
+};
+
+class ProcessOrder {
+  // Hàm để xử lý đơn hàng
+  @Transactional()
+  static async processOrder(req: Request, res: Response, imagePath: string) {
+    const user = await checkLoggedIn(req, res);
+    const { name, phone, address, typeOrder, totalAmount } = req.body;
+    const order = await createOrder(name, phone, address, typeOrder, totalAmount, imagePath, user);
+    await orderRepository.save(order);
+
+    const cart = await getUserCart(user.id, res);
+    const cartItems = await getCartItems(cart);
+
+    await createOrderDetails(order, cartItems);
+    await updateProductQuantities(cartItems);
+    await deleteCartItems(cart);
+
+    runOnTransactionRollback(() => console.log('rollbacked'));
+    runOnTransactionCommit(() => console.log('committed'));
+
+    res.redirect('/order/success');
+  }
+}
+
+// Hàm để tạo đơn hàng
+const createOrder = async (name: string, phone: string, address: string, typeOrder: number, totalAmount, imagePath: string, user: User) => {
```

Comment:
e truyền tham số dạng options arg như này cho ngắn gọn nhé, lúc gọi thì truyền đúng keys là đc e
```suggestion
const createOrder = async (user: User, options={}) => {
```